receiver/vcenterreceiver/metrics.go

```diff
@@ -6,11 +6,77 @@ package vcenterreceiver // import "github.com/open-telemetry/opentelemetry-colle
 	"go.opentelemetry.io/collector/pdata/pcommon"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver/internal/metadata"
 )
 
+// recordDatacenterStats records stat metrics for a vSphere Datacenter
+func (v *vcenterMetricScraper) recordDatacenterStats(
+	ts pcommon.Timestamp,
+	dcStat *DatacenterStats,
+) {
+	// Cluster metrics
+	v.mb.RecordVcenterDatacenterClusterCountDataPoint(ts, dcStat.ClusterStatus[types.ManagedEntityStatusRed], metadata.AttributeEntityStatusRed)
+	v.mb.RecordVcenterDatacenterClusterCountDataPoint(ts, dcStat.ClusterStatus[types.ManagedEntityStatusYellow], metadata.AttributeEntityStatusYellow)
+	v.mb.RecordVcenterDatacenterClusterCountDataPoint(ts, dcStat.ClusterStatus[types.ManagedEntityStatusGreen], metadata.AttributeEntityStatusGreen)
+	v.mb.RecordVcenterDatacenterClusterCountDataPoint(ts, dcStat.ClusterStatus[types.ManagedEntityStatusGray], metadata.AttributeEntityStatusGray)
+
+	// datastore metrics
+	v.mb.RecordVcenterDatacenterDatastoreCountDataPoint(ts, dcStat.DatastoreCount)
+
+	var entityStatusToAttribute = map[types.ManagedEntityStatus]metadata.AttributeEntityStatus{
+		types.ManagedEntityStatusRed:    metadata.AttributeEntityStatusRed,
+		types.ManagedEntityStatusYellow: metadata.AttributeEntityStatusYellow,
+		types.ManagedEntityStatusGreen:  metadata.AttributeEntityStatusGreen,
+		types.ManagedEntityStatusGray:   metadata.AttributeEntityStatusGray,
+	}
+
+	var vmPowerStateToAttribute = map[string]metadata.AttributeVMCountPowerState{
+		"poweredOn":  metadata.AttributeVMCountPowerStateOn,
+		"poweredOff": metadata.AttributeVMCountPowerStateOff,
+		"suspended":  metadata.AttributeVMCountPowerStateSuspended,
+	}
+
+	var hostPowerStateToAttribute = map[string]metadata.AttributeHostPowerState{
+		"poweredOn":  metadata.AttributeHostPowerStateOn,
+		"poweredOff": metadata.AttributeHostPowerStateOff,
+		"standby":    metadata.AttributeHostPowerStateStandby,
+		"unknown":    metadata.AttributeHostPowerStateUnknown,
+	}
+
+	// VM metrics
+	for powerState, vmInfo := range dcStat.VMStats {
+		for status, count := range vmInfo {
+			entityStatus, errStatus := entityStatusToAttribute[status]
+			vmPowerState, errPowerState := vmPowerStateToAttribute[powerState]
+			if !errStatus || !errPowerState {
+				continue
+			}
+			v.mb.RecordVcenterDatacenterVMCountDataPoint(ts, count, entityStatus, vmPowerState)
+		}
+	}
+
+	// Host metrics
+	for powerState, hostInfo := range dcStat.HostStats {
+		for status, count := range hostInfo {
+			entityStatus, errStatus := entityStatusToAttribute[status]
+			hostPowerState, errPowerState := hostPowerStateToAttribute[powerState]
+			if !errStatus || !errPowerState {
+				continue
+			}
+			v.mb.RecordVcenterDatacenterHostCountDataPoint(ts, count, entityStatus, hostPowerState)
+		}
+	}
+
+	// Datacenter stats
+	v.mb.RecordVcenterDatacenterDiskSpaceDataPoint(ts, (dcStat.DiskCapacity - dcStat.DiskFree), metadata.AttributeDiskStateUsed)
```

Comment:
I was actually initially imagining this metric to ONLY be for a total disk size....but this actually works better I think. Ignore my comments for the metadata.yaml, and instead just make sure the descriptions generally match what is currently used for the similar datastore metric.

Reply:
resolve